### DISCUSSION TITLE
A way to detect a composable is being used in setup store

### DISCUSSION THREAD

**Discussion by MartinMalinda** [vuejs/pinia#2713] [Help and Questions] [answered]
### What problem is this solving

In [vue-concurrency ](https://github.com/MartinMalinda/vue-concurrency) I have a composable called "useTask" and most common usecase for it is to run it in components. But there's also usecase for global tasks and pinia is a good fit.

Currently I issue a warning in the console when a task is being used outside of component setup. Which is annoying if someone uses it in pinia.

It would be great if I, as a library author, could detect that the composable is being run inside pinia. I could then turn of this warning.

### Proposed solution

I can imagine I could use an equivalent to `getCurrentInstance` as I already use `getCurrentInstance` to detect if the task is correctly created inside setup.

So something like `getCurrentStore()` although it could be confused with `getActivePinia()`.

Alternatively I could just look up some flag on some global object? This would also mean I don't have to have `pinia` as dependency.

### Describe alternatives you've considered

I can provide an API to explicitly turn off the warnings but it would be great if I didn't have to do this.

**Comment by posva** [comment, marked as answer]
You can use [`getCurrentScope()`](https://vuejs.org/api/reactivity-advanced.html#getcurrentscope) to know if you are within an effect scope.

You shouldn't need `pinia` as a dependency, that's for sure. In library code, you can check for a currentInstance and conditionally setup hooks if needed but for `onUnmounted()` you can also use `onScopeDispose()`. This should make your library work with libraries other than Pinia that use the same pattern

**Reply by MartinMalinda** [reply]
that's perfect, I didn't know about this function, thank you! 🙏

**Comment by victorgarciaesgi** [comment]
Hey @MartinMalinda @posva ! Could you elaborate on how to detect with `getCurrentScope` that it's being ran inside a Pinia store?

I have a case on https://github.com/victorgarciaesgi/regle where I want to use `skipHydrate` on the main stateful property but only if it's ran inside a Pinia store.

**Reply by victorgarciaesgi** [reply]
Right now `getCurrentScope` returns nothing but a list of effects, and nothing can indicate it's a Pinia store

**Reply by victorgarciaesgi** [reply]
Also, `getActivePinia` is always undefined

**Reply by MartinMalinda** [reply]
In my case I actually didn't need to detect pinia specifically, I just wanted to warn the user not to create tasks outside of active scope. So whether it's compoent or pinia it does not matter. I would just cancel the task if the scope it was created in got inactivated.

I guess you need to specifically check for Pinia because Pinia hydrates on its own?

**Reply by MartinMalinda** [reply]
Later on I even reverted `getCurrentScope` and `onScopeDispose` usage because it caused some problems with reactivity on my end (https://github.com/MartinMalinda/vue-concurrency/issues/107)

**Reply by victorgarciaesgi** [reply]
Yep saw your changes 😅 
On my side it's more a case of SSR and using `store.$dispose()`. 
The easiest way for now is to have a section in my docs for Pinia to advise to use `skipHydrate` if I can't do it automatically
